File: app/src/main/java/com/amaury/pointage/v2/engine/EmployerApprenticeshipTaxV2.kt
package com.amaury.pointage.v2.engine

import java.time.YearMonth

/**
 * Taxe d'apprentissage employeur rattachée au coût mensuel de la rémunération.
 * La part principale et le solde sont séparés car leur calendrier de paiement diffère.
 * Les taux sont confirmés par entreprise/période : aucune exonération ni régime local n'est deviné.
 */
object EmployerApprenticeshipTaxV2 {
    data class Record(
        val id: String,
        /** Taux décimal, ex. 0,59 % = 0,0059 ; 0,44 % = 0,0044. */
        val principalRate: Double,
        /** Provision économique mensuelle du solde, ex. 0,09 % = 0,0009 ; 0 en Alsace-Moselle/exonération. */
        val balanceRate: Double,
        val effectiveFrom: YearMonth,
        val effectiveTo: YearMonth? = null,
        val source: String
    )

    data class Snapshot(
        val principalRate: Double?,
        val balanceRate: Double?,
        val source: String?,
        val reliable: Boolean,
        val warnings: List<String>
    )

    data class Result(
        val principalAmount: Double?,
        val balanceAccrualAmount: Double?,
        val totalEmployerAmount: Double?,
        val complete: Boolean,
        val warnings: List<String>
    )

    fun resolve(records: List<Record>, period: YearMonth): Snapshot {
        val malformed = records.filter {
            it.source.isBlank() ||
                it.effectiveTo?.let { end -> end < it.effectiveFrom } == true ||
                !validRate(it.principalRate) || !validRate(it.balanceRate)
        }
        if (malformed.isNotEmpty()) {
            return Snapshot(null,null,null,false,listOf("Taxe d’apprentissage : une règle est incomplète ou incohérente ; aucun montant patronal n'est appliqué."))
        }
        val active = records.filter {
            period >= it.effectiveFrom && (it.effectiveTo == null || period <= it.effectiveTo)
        }
        if (active.isEmpty()) {
            return Snapshot(null,null,null,false,listOf("Taxe d’apprentissage : taux/applicabilité à confirmer pour ${period.monthValue.toString().padStart(2,'0')}/${period.year}."))
        }
        if (active.size > 1) {
            return Snapshot(null,null,null,false,listOf("Taxe d’apprentissage : plusieurs règles se chevauchent ; calcul patronal bloqué."))
        }
        val selected=active.single()
        return Snapshot(selected.principalRate,selected.balanceRate,selected.source,true,emptyList())
    }

    fun calculate(grossSocial: Double, principalRate: Double?, balanceRate: Double?): Result {
        if (principalRate == null || balanceRate == null) {
            return Result(null,null,null,false,listOf("Taxe d’apprentissage : taux confirmés manquants ; coût employeur incomplet."))
        }
        if (!validRate(principalRate) || !validRate(balanceRate)) {
            return Result(null,null,null,false,listOf("Taxe d’apprentissage : taux invalide ; aucun montant n'est calculé."))
        }
        val base=grossSocial.coerceAtLeast(0.0)
        val principal=base*principalRate
        val balance=base*balanceRate
        return Result(principal,balance,principal+balance,true,emptyList())
    }

    private fun validRate(rate: Double): Boolean = rate.isFinite() && rate >= 0.0 && rate <= 1.0
}

File: ios/HPTravail/HPTravail/SalaryV2/EmployerApprenticeshipTaxV2.swift
import Foundation

/// Taxe d'apprentissage employeur rattachée au coût mensuel de la rémunération.
/// La part principale et le solde sont séparés et les taux doivent être confirmés par entreprise/période.
enum EmployerApprenticeshipTaxV2 {
    struct Period: Equatable, Comparable {
        let year: Int
        let month: Int

        var isValid: Bool { year > 0 && (1...12).contains(month) }

        static func < (lhs: Period, rhs: Period) -> Bool {
            lhs.year == rhs.year ? lhs.month < rhs.month : lhs.year < rhs.year
        }

        var label: String { String(format: "%02d/%04d", month, year) }
    }

    struct Record: Equatable {
        let id: String
        let principalRate: Double
        let balanceRate: Double
        let effectiveFrom: Period
        let effectiveTo: Period?
        let source: String
    }

    struct Snapshot: Equatable {
        let principalRate: Double?
        let balanceRate: Double?
        let source: String?
        let reliable: Bool
        let warnings: [String]
    }

    struct Result: Equatable {
        let principalAmount: Double?
        let balanceAccrualAmount: Double?
        let totalEmployerAmount: Double?
        let complete: Bool
        let warnings: [String]
    }

    static func resolve(records: [Record], period: Period) -> Snapshot {
        guard period.isValid else {
            return unresolved("Taxe d’apprentissage : période invalide ; aucun montant patronal n'est appliqué.")
        }

        let malformed = records.filter { record in
            record.source.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
                !record.effectiveFrom.isValid ||
                (record.effectiveTo?.isValid == false) ||
                (record.effectiveTo.map { $0 < record.effectiveFrom } ?? false) ||
                !validRate(record.principalRate) ||
                !validRate(record.balanceRate)
        }
        guard malformed.isEmpty else {
            return unresolved("Taxe d’apprentissage : une règle est incomplète ou incohérente ; aucun montant patronal n'est appliqué.")
        }

        let active = records.filter { record in
            period >= record.effectiveFrom && (record.effectiveTo == nil || period <= record.effectiveTo!)
        }
        guard !active.isEmpty else {
            return unresolved("Taxe d’apprentissage : taux/applicabilité à confirmer pour \(period.label).")
        }
        guard active.count == 1, let selected = active.first else {
            return unresolved("Taxe d’apprentissage : plusieurs règles se chevauchent ; calcul patronal bloqué.")
        }

        return Snapshot(
            principalRate: selected.principalRate,
            balanceRate: selected.balanceRate,
            source: selected.source,
            reliable: true,
            warnings: []
        )
    }

    static func calculate(
        grossSocial: Double,
        principalRate: Double?,
        balanceRate: Double?
    ) -> Result {
        guard let principalRate, let balanceRate else {
            return incomplete("Taxe d’apprentissage : taux confirmés manquants ; coût employeur incomplet.")
        }
        guard validRate(principalRate), validRate(balanceRate) else {
            return incomplete("Taxe d’apprentissage : taux invalide ; aucun montant n'est calculé.")
        }
        guard grossSocial.isFinite, grossSocial >= 0 else {
            return incomplete("Taxe d’apprentissage : assiette brute sociale invalide ; aucun montant patronal n'est calculé.")
        }

        let principal = grossSocial * principalRate
        let balance = grossSocial * balanceRate
        return Result(
            principalAmount: principal,
            balanceAccrualAmount: balance,
            totalEmployerAmount: principal + balance,
            complete: true,
            warnings: []
        )
    }

    private static func unresolved(_ warning: String) -> Snapshot {
        Snapshot(
            principalRate: nil,
            balanceRate: nil,
            source: nil,
            reliable: false,
            warnings: [warning]
        )
    }

    private static func incomplete(_ warning: String) -> Result {
        Result(
            principalAmount: nil,
            balanceAccrualAmount: nil,
            totalEmployerAmount: nil,
            complete: false,
            warnings: [warning]
        )
    }

    private static func validRate(_ rate: Double) -> Bool {
        rate.isFinite && rate >= 0 && rate <= 1
    }
}
